propagators/opentelemetry-propagator-instana/src/InstanaPropagator.ts

```diff
@@ -0,0 +1,125 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Context,
+  trace,
+  isSpanContextValid,
+  isValidSpanId,
+  isValidTraceId,
+  TextMapGetter,
+  TextMapPropagator,
+  TextMapSetter,
+  TraceFlags,
+  INVALID_TRACEID,
+  INVALID_SPANID,
+} from '@opentelemetry/api';
+
+/** Instana header names */
+export const INSTANA_TRACE_ID_HEADER = 'X-INSTANA-T';
+export const INSTANA_SPAN_ID_HEADER = 'X-INSTANA-S';
+export const INSTANA_LEVEL_HEADER = 'X-INSTANA-L';
+
+const FIELDS = [
+  INSTANA_TRACE_ID_HEADER,
+  INSTANA_SPAN_ID_HEADER,
+  INSTANA_LEVEL_HEADER,
+];
+
+function readHeader(
+  carrier: unknown,
+  getter: TextMapGetter,
+  key: string
+): string {
+  // By convention, X-INSTANA-* headers are sent all-upper-case. For http and http/2, Node.js normalizes all headers to
+  // all-lower-case, that's why we first read via the lower case variant of the header name. For other protocols, no
+  // such normalization happens, so we also need to read via the original all-upper-case variant.
+  let header =
+    getter.get(carrier, key.toLowerCase()) || getter.get(carrier, key);
+  if (Array.isArray(header)) [header] = header;
+  return header || '';
+}
+
+/**
+ * Propagator for the Instana HTTP headers.
+ */
+export class InstanaPropagator implements TextMapPropagator {
+  /**
+   * Injects the current span context into Instana's vendor specific trace correlation headers (X-INSTANA-T, X-INSTANA-S
+   * and X-INSTANA-L).
+   */
+  inject(context: Context, carrier: unknown, setter: TextMapSetter) {
+    const spanContext = trace.getSpan(context)?.spanContext();
+    if (!spanContext || !isSpanContextValid(spanContext)) {
+      return;
+    }
+
+    setter.set(carrier, INSTANA_TRACE_ID_HEADER, spanContext.traceId);
+    setter.set(carrier, INSTANA_SPAN_ID_HEADER, spanContext.spanId);
+    const sampled =
+      (spanContext.traceFlags & TraceFlags.SAMPLED) == TraceFlags.SAMPLED;
+    setter.set(carrier, INSTANA_LEVEL_HEADER, sampled ? '1' : '0');
+  }
+
+  /**
+   * Extracts the span context from Instana's vendor specific trace correlation headers (X-INSTANA-T, X-INSTANA-S
+   * and X-INSTANA-L).
+   */
+  extract(context: Context, carrier: unknown, getter: TextMapGetter): Context {
+    let traceId = readHeader(carrier, getter, INSTANA_TRACE_ID_HEADER);
+    if (traceId && traceId.length < 32) {
+      traceId = traceId.padStart(32, '0');
+    }
+    let spanId = readHeader(carrier, getter, INSTANA_SPAN_ID_HEADER);
+    if (spanId && spanId.length < 16) {
+      spanId = spanId.padStart(16, '0');
+    }
+    const level = readHeader(carrier, getter, INSTANA_LEVEL_HEADER) || '1';
+    const sampled = level !== '0';
+    const traceFlags = sampled ? TraceFlags.SAMPLED : TraceFlags.NONE;
+
+    if (isValidTraceId(traceId) && isValidSpanId(spanId)) {
```

Comment:
@basti1302 There is one more edge case we might want to handle.

The services sends `X-INSTANA-L=1`, but does not send the trace & span ID. This is not a realistic case,, but can happen. Wondering if we want to create a context in this case (with a invalid trace ID, spanID and the sampled flag). Right now there won't be any new context information.

Reply:
I'm -1 on handling that case explicitly. First of all it is invalid behavior by the upstream Instana tracer, plus, AFAICT, it does no provide any benefit to consumers of the context.